contracts/QuestNFT.sol

```diff
@@ -0,0 +1,219 @@
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        uint256 endTime_,
+        uint256 startTime_,
+        uint256 totalParticipants_,
+        string memory questId_,
+        uint256 questFee_,
+        address protocolFeeRecipient_,
+        address minterAddress_, // should always be the QuestFactory contract
+        string memory jsonSpecCID_,
+        string memory name_,
+        string memory symbol_,
+        string memory description_,
+        string memory imageIPFSHash_
+    ) external initializer {
+        require (endTime_ > block.timestamp, 'endTime_ in past');
+        require (startTime_ > block.timestamp, 'startTime_ in past');
+        require (endTime_ > startTime_, 'startTime_ before endTime_');
+        endTime = endTime_;
+        startTime = startTime_;
+        totalParticipants = totalParticipants_;
+        questId = questId_;
+        questFee = questFee_;
+        protocolFeeRecipient = protocolFeeRecipient_;
+        jsonSpecCID = jsonSpecCID_;
+        minterAddress = minterAddress_;
+        imageIPFSHash = imageIPFSHash_;
+        description = description_;
+        __ERC721_init(name_, symbol_);
+        __Ownable_init();
+        __Pausable_init();
+        __ReentrancyGuard_init();
+    }
+
+    /// @notice Prevents reward withdrawal until the Quest has ended
+    modifier onlyAfterQuestEnd() {
+        require (block.timestamp > endTime, 'Quest has not ended');
+        _;
+    }
+
+    /// @notice Checks if quest has started from the start time
+    modifier onlyQuestBetweenStartEnd() {
+        require(block.timestamp > startTime, 'Quest not started');
+        require(block.timestamp < endTime, 'Quest ended');
+        _;
+    }
+
+    modifier onlyMinter() {
+        require(msg.sender == minterAddress, 'Only minter address');
+        _;
+    }
+
+    /// @dev set jsonSpecCID only if its empty
+    /// @param jsonSpecCID_ The jsonSpecCID to set
+    function setJsonSpecCID(string memory jsonSpecCID_) external onlyOwner {
+        require(bytes(jsonSpecCID_).length > 0, 'jsonSpecCID cannot be empty');
+        require(bytes(jsonSpecCID).length == 0, 'jsonSpecCID already set');
+
+        jsonSpecCID = jsonSpecCID_;
+        emit JsonSpecCIDSet(jsonSpecCID_);
+    }
+
+    /// @dev The maximum amount of coins the quest needs for the protocol fee
+    function totalTransferAmount() external view returns (uint256) {
+        return questFee * totalParticipants;
+    }
+
+    /// @notice Pauses the Quest
+    /// @dev Only the owner of the Quest can call this function.
+    function pause() external onlyOwner {
+        _pause();
+    }
+
+    /// @notice Unpauses the Quest
+    /// @dev Only the owner of the Quest can call this function.
+    function unPause() external onlyOwner {
+        _unpause();
+    }
+
+    function safeMint(address to_) public onlyMinter onlyQuestBetweenStartEnd whenNotPaused nonReentrant {
```

Comment:
Do we want to cap the amount of mints? Maybe to the value of `totalParticipants`?